src/rdf4cpp/rdf/datatypes/registry/DatatypeConversion.hpp

```diff
@@ -0,0 +1,355 @@
+ *
+ * @tparam BaseType the type for which to generate the linear conversions
+ * @tparam Type the current type getting processed in the linear hierarchy
+ * @tparam Conversion a common interface to carry out a conversion (see adaptor namespace)
+ * @tparam HasConversion check if a LiteralDatatype has a next conversion, e.g. if it is promotable
+ * @tparam RankRule a boolean predicate to check if the defined conversion is valid
+ * @tparam ConvertEntries types of the already existing ConversionEntries
+ * @param table_acc accumulator for ConversionEntries
+ * @return a std::tuple<ConversionEntry, ...> containing all linear conversions reachable from BaseType
+ */
+template<LiteralDatatype BaseType,
+         LiteralDatatype Type,
+         template<LiteralDatatype> typename Conversion,
+         template<LiteralDatatype> typename HasConversion,
+         template<LiteralDatatype, LiteralDatatype> typename RankRule,
+         ConversionLayer LayerAcc>
+consteval ConversionLayer auto make_conversion_layer_impl(LayerAcc const &table_acc) {
+    if constexpr (!HasConversion<Type>::value) {
+        return table_acc;
+    } else {
+        using next = Conversion<Type>;
+
+        // defining a hierarchy which does not follow the rank rules (see adaptors::PromoteRankRule, adaptors::SubtypeRankRule)
+        // would result in not being able to find all conversions between types (because of assumptions used in the search algorithm)
+        // and thus yield unexpected behaviour when trying to convert
+        // e.g. with
+        //        A --> Z               Legend:
+        //        |     |                   --> : promotable to
+        //        B --> Y --> X (!)          |  : subtype of
+        //
+        // The algorithm would not be able to find B --> X or Y --> X,
+        // because X has subtype rank 0 while Y and B have subtype rank 1.
+        // This is problematic when trying to jump to the common subtype level for both arguments
+        // because while in the graph they are visually at the correct level, their subtype ranks do not reflect that.
+        // The same problem exists when a supertype (or promoted type) has a subtype (promotion) rank that is not exactly one less than the origin type's rank.
+        static_assert(RankRule<Type, typename next::converted>::value,
+                      "detected invalid hierarchy, would not be able to discover all conversions");
+
+        // conversion must preserve numericity, so:
+        // NumericLiteralDatatype<Type> must imply NumericLiteralDatatype<typename next::converted>
+        // This is because the numeric operations assume that if any of the operands is not numeric
+        // the common type and the result type will also not be numeric to allow for an early return
+        // without doing the actual conversion.
+        static_assert(!NumericLiteralDatatype<Type> || NumericLiteralDatatype<typename next::converted>,
```

Comment:
Technically this assertion is not strong enough because it means even if the result would be comparable
if one of the source types is not we do an early (null) return in the literal operations, which is potentially unexpected behaviour.

Should I change the one-sided implication to a two sided one? (So from -> to <->)
I mean it is technically not required for it to be two sided but maybe the behaviour would be more consistent(?).

Please share your thoughts.

Reply:
I would say better be too strict than having undefined behavior.